### DISCUSSION TITLE
Documentation for migrating v2 to v3

### DISCUSSION THREAD

**Discussion by johnwc** [elsa-workflows/elsa-core#4767] [Q&A]
Is there documentation to help in migrating existing v2 workflows to v3? As well as migrating custom activities from v2 to v3? Ex: how do we create a long running activity in v3? You mentioned previously that we should be using hangfire in v2, but in v3 it would be built into the engine and not needed.

**Comment by johnwc** [comment]
@sfmskywalker Looking forward to starting the migration

**Comment by sfmskywalker** [comment]
Hi @johnwc,

Thanks for your question regarding the use of Hangfire in Elsa V3.

I've just created an issue in the [elsa-documentation](https://github.com/elsa-workflows/elsa-documentation) repo to address this topic in detail, which you can track here: [Elsa Documentation Issue #50](https://github.com/elsa-workflows/elsa-documentation/issues/50). Consequently, I'll close this discussion to centralise our conversation there.

To give you a quick overview:

In V3, we've introduced a default background activity scheduler (`IBackgroundActivityScheduler`) built on a hosted service and the .NET Channels API. This is a pretty simple implementation and does not implement aspects like job failover and retries. For those more advanced capabilities, it is recommend to use the Hangfire implementation.

Here's a snapshot of the current implementations of `IBackgroundActivityScheduler`:

- LocalBackgroundActivityScheduler
- HangfireBackgroundActivityScheduler

Another advantage of the Hangfire implementation is that background work can now be distributed across multiple nodes in your cluster.

It's also possible to integrate other systems like MassTransit for background job execution by implementing the `IBackgroundActivityScheduler` interface if you prefer to use something else than Hangfire.

Regarding custom long-running activities, here's how you can define one:

```csharp
[Activity(Kind = ActivityKind.Job)]
public class ProcessRawVideo : CodeActivity
{
    public override async Task ExecuteAsync(ActivityExecutionContext context)
    {
        // Your video processing logic. As a Job type activity, it'll be scheduled for execution in the background.
    }
}
```

This code snippet illustrates how to define an activity that processes tasks in the background, allowing the workflow to continue processing other activities while waiting for completion of the background activity.

For a more in-depth understanding, I recommend this [blog post](https://medium.com/@sipkeschoorstra/parallel-task-orchestration-with-elsa-3-0-e387f863655f) by yours truly, which delves into parallel task orchestration with Elsa 3.0.

Feel free to ask if you have more questions or need further clarifications!

**Reply by johnwc** [reply]
Great! I will start looking into this and read over your blog post. Is the Actor functionality only for executing workflows, I remember reading that with the new use of this library that all the locking is removed and supports more robust execution in a multi-node environment. I was hoping this was the case for background jobs as well?

**Comment by sfmskywalker** [comment]
But wait, this is a discussion and not an issue, so no need to close this out necessarily 😄
